src/miniscript/context.rs

```diff
@@ -0,0 +1,200 @@
+    /// estimate the cost of satisfaction because of compressed keys
+    /// This is currently only used in compiler code for removing malleable
+    /// compilations.
+    /// This does NOT recursively check if the children of the fragment are
+    /// valid or not. Since the compilation proceeds in a leaf to root fashion,
+    /// a recursive check is unnecessary.
+    fn check_frag_non_malleable<Pk: MiniscriptKey, Ctx: ScriptContext>(
+        _frag: &Terminal<Pk, Ctx>,
+    ) -> Result<(), ScriptContextError>;
+
+    /// Depending on script Context, some of the Terminals might not be valid.
+    /// For example, in Segwit Context with MiniscriptKey as bitcoin::PublicKey
+    /// uncompressed public keys are non-standard and thus invalid.
+    /// Post Tapscript upgrade, this would have to consider other nodes.
+    /// This does not recursively check
+    fn check_frag_validity<Pk: MiniscriptKey, Ctx: ScriptContext>(
+        _frag: &Terminal<Pk, Ctx>,
+    ) -> Result<(), ScriptContextError>;
+}
+
+/// Legacy ScriptContext
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub enum Legacy {}
+
+impl ScriptContext for Legacy {
+    fn check_frag_non_malleable<Pk: MiniscriptKey, Ctx: ScriptContext>(
+        frag: &Terminal<Pk, Ctx>,
+    ) -> Result<(), ScriptContextError> {
+        match *frag {
+            Terminal::PkH(ref _pkh) => Err(ScriptContextError::MalleablePkH),
+            Terminal::OrI(ref _a, ref _b) => Err(ScriptContextError::MalleableOrI),
+            Terminal::DupIf(ref _ms) => Err(ScriptContextError::MalleableDupIf),
+            _ => Ok(()),
+        }
+    }
+
+    fn check_frag_validity<Pk: MiniscriptKey, Ctx: ScriptContext>(
+        _frag: &Terminal<Pk, Ctx>,
+    ) -> Result<(), ScriptContextError> {
+        Ok(())
+    }
+}
+
+/// Segwitv0 ScriptContext
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub enum Segwitv0 {}
+
+impl ScriptContext for Segwitv0 {
+    fn check_frag_non_malleable<Pk: MiniscriptKey, Ctx: ScriptContext>(
+        _frag: &Terminal<Pk, Ctx>,
+    ) -> Result<(), ScriptContextError> {
+        Ok(())
+    }
+
+    fn check_frag_validity<Pk: MiniscriptKey, Ctx: ScriptContext>(
+        frag: &Terminal<Pk, Ctx>,
+    ) -> Result<(), ScriptContextError> {
+        match *frag {
+            Terminal::PkK(ref pk) if pk.is_uncompressed() => {
+                Err(ScriptContextError::CompressedOnly)
+            }
+            _ => Ok(()),
+        }
+    }
+}
+
+/// Any ScriptContext. The implementation has the most conservative rules
+/// for validity and malleability
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub enum Any {}
+impl ScriptContext for Any {
+    fn check_frag_non_malleable<Pk: MiniscriptKey, Ctx: ScriptContext>(
+        frag: &Terminal<Pk, Ctx>,
+    ) -> Result<(), ScriptContextError> {
+        Segwitv0::check_frag_non_malleable(frag)?;
+        Legacy::check_frag_non_malleable(frag)
```

Comment:
I think these should be `unimplemented!()` rather than doing both checks. We shouldn't ever actually use `Any` except in places where the context doesn't matter.

Reply:
Changed to `!unreachable!()`